### PR TITLE
[frameworks] Convert to TypeScript

### PR DESCRIPTION
**Pull request by TooTallNate**
This PR converts the `frameworks.json` file to TypeScript, and extends the values with the detection logic from `@vercel/static-build`, so that it's publicly editable. You also don't need to do the type casting downstream anymore.

As a consequence, it also makes Zola a 1st-class framework, as it was previously missing from the `frameworks.json` file, but present in the static-build frameworks. An example has been included based on their "Getting Started" tutorial.

CH-3808
CH-18771

### CODE REVIEW

File: packages/frameworks/package.json [hunk 1]
```diff
@@ -1,14 +1,22 @@
 {
   "name": "@vercel/frameworks",
   "version": "0.2.0",
-  "main": "frameworks.json",
+  "main": "./dist/frameworks.js",
+  "types": "./dist/frameworks.d.ts",
   "license": "UNLICENSED",
```
Comment (styfle) [inline packages/frameworks/package.json:9]: You should add `"files": [ "dist" ]` to avoid publishing unnecessary files to npm

File: packages/frameworks/test/frameworks.unit.test.ts [hunk 2]
```diff
@@ -54,7 +53,6 @@ const Schema = {
   items: {
     type: 'object',
     required: ['name', 'slug', 'logo', 'description', 'settings'],
```
Comment (styfle) [inline packages/frameworks/test/frameworks.unit.test.ts]: Let's add the additional required props to this test

File: utils/run.js [hunk 3]
```diff
@@ -3,8 +3,8 @@ const { join, relative, sep } = require('path');
 
 // The order matters because we must build dependencies first
 const allPackages = [
-  'frameworks',
   'now-routing-utils',
+  'frameworks',
```
Comment (styfle) [inline utils/run.js:7]: Why the change of order here?

Comment (styfle) [reply utils/run.js:7]: Now that I think about it, maybe frameworks should be built after build-utils so it can import `readConfigFile`

File: packages/frameworks/src/frameworks.ts [hunk 4]
```diff
@@ -0,0 +1,1605 @@
...
+      outputDirectory: {
+        value: 'build',
+      },
+    },
+    devCommand: 'bundle exec middleman server -p $PORT',
+    buildCommand: 'bundle exec middleman build',
+    getOutputDirName: async () => 'build',
+  },
+  {
+    name: 'RedwoodJS',
```
Comment (styfle) [inline packages/frameworks/src/frameworks.ts]: Any reason why you changed the order of RedwoodJS and Docusaurus 1?

File: packages/frameworks/src/types.ts [hunk 5]
```diff
@@ -1,3 +1,5 @@
+import { Route } from '@vercel/routing-utils/dist/types';
```
Comment (styfle) [inline packages/frameworks/src/types.ts]:
```suggestion
import { Route } from '@vercel/routing-utils';
```

File: utils/run.js (hunk 3 above)
Comment (TooTallNate) [reply utils/run.js:7]: Cause `frameworks` requires the `Route` type from `routing-utils` (see src/types.ts).

File: packages/frameworks/src/types.ts (hunk 5 above)
Comment (TooTallNate) [reply packages/frameworks/src/types.ts]: I did it like this so that downstream users that only load the types (i.e. front only needs `Framework` type) don't have to load all of routing-utils into memory.

File: packages/frameworks/src/frameworks.ts (hunk 4 above)
Comment (TooTallNate) [reply packages/frameworks/src/frameworks.ts]: Accidental, I'll fix.

File: packages/frameworks/test/frameworks.unit.test.ts (hunk 2 above)
Comment (TooTallNate) [reply packages/frameworks/test/frameworks.unit.test.ts]: I'll add the string ones, but there's no way to test functions with `ajv` so maybe we can manually check for that after ajv.

File: packages/frameworks/src/types.ts (hunk 5 above)
Comment (styfle) [reply packages/frameworks/src/types.ts]: Won't it be stripped at build time so it wont be emitted in the bundle js in front?

Also you could try: `import type { Route } from '@vercel/routing-utils';`

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export

File: utils/run.js (hunk 3 above)
Comment (styfle) [reply utils/run.js:7]: Maybe we should remove `readConfigFile` from build-utils and anything relying on it can import from frameworks instead?

Comment (TooTallNate) [reply utils/run.js:7]: That's what this PR is doing 😛

File: packages/frameworks/src/types.ts (hunk 5 above)
Comment (TooTallNate) [reply packages/frameworks/src/types.ts]: Oh ya I think you're right, that's a good point.

**Review by styfle — CHANGES_REQUESTED** [review]
The api is failing with the following:

```
ERROR	Cannot find module '../packages/frameworks'
Require stack:
- /var/task/api/frameworks.ts
```

Also there are build errors: https://vercel.com/vercel/now/f25ig4pt5

We should enable `noEmitOnError: true` in tsconfig.json

**Comment by TooTallNate** [comment]
Nice catch

File: api/package.json [hunk 6]
```diff
@@ -5,7 +5,7 @@
   "description": "API for the vercel/vercel repo",
   "main": "index.js",
   "scripts": {
-    "build": "yarn --cwd .. && node ../utils/run.js build all"
+    "vercel-build": "yarn --cwd .. && node ../utils/run.js build all"
```
Comment (styfle) [inline api/package.json:8]: I think the API should probably strip the new properties so it remains unchanged. Compare your PR deployment to https://vercel.com/api/v1/frameworks and you'll see the diff.

File: packages/frameworks/src/frameworks.ts [hunk 7]
```diff
@@ -0,0 +1,1605 @@
...
+        continue: true,
+      },
+      {
+        src: '/(.*)',
+        dest: '/index.html',
+      },
+    ],
+  },
+  {
+    name: 'Ember',
```
Comment (styfle) [inline packages/frameworks/src/frameworks.ts]:
```suggestion
    name: 'Ember.js',
```

Was this intentionally changed?

File: packages/frameworks/src/frameworks.ts [hunk 8]
```diff
@@ -0,0 +1,1605 @@
...
+      },
+      {
+        src: '.*',
+        status: 404,
+        dest: '404.html',
+      },
+    ],
+  },
+  {
+    name: 'Docusaurus 1.0',
```
Comment (styfle) [inline packages/frameworks/src/frameworks.ts]:
```suggestion
    name: 'Docusaurus 1',
```

Was this intentionally changed?

File: api/package.json (hunk 6 above)
Comment (TooTallNate) [reply api/package.json:8]: 8830dd9f559685a9b42156629dd2a90d5495391c

File: packages/frameworks/src/frameworks.ts (hunk 7 above)
Comment (TooTallNate) [reply packages/frameworks/src/frameworks.ts]: Updated, but to answer your question, these were the values from `static-build`.

File: packages/frameworks/src/frameworks.ts [hunk 9]
```diff
@@ -0,0 +1,1605 @@
...
+      {
+        src: '/(.*)',
+        headers: { 'cache-control': 's-maxage=0' },
+        dest: '/index.html',
+      },
+    ],
+  },
+  {
+    name: 'Create React App (ejected)',
+    slug: 'create-react-app',
```
Comment (styfle) [inline packages/frameworks/src/frameworks.ts]: Why are there two entries for `create-react-app`? I think you can remove this one.

**Review by styfle — DISMISSED** [review]
Looks good, thanks! 🏅

**Review by styfle — APPROVED** [review]

**Review by AndyBitz — APPROVED** [review]
